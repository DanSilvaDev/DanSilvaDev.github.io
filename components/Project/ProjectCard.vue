<template>
  <v-card outlined>
    <v-container fluid class="fill-height">
      <v-row>
        <slot name="header">
          <v-col cols="12" align="center">
            <h2>{{ title }}</h2>
          </v-col>
        </slot>
      </v-row>
      <v-row justify="center" align="center">
        <slot name="default">
          <v-col
            cols="12"
            align="center"
            justify="center"
          >
            <div class="d-flex align-stretch">
              <v-img
                height="100%"
                max-height="150px"
                :src="imgSrc"
              />
            </div>
          </v-col>
          <v-col
            cols="12"
            align="center"
          >
            <p class="text-justify">
              {{ description }}
            </p>
          </v-col>
        </slot>
      </v-row>
      <v-row>
        <v-divider />
      </v-row>
      <v-row>
        <slot name="footer">
          <v-col cols="12" align="right">
            <v-container fluid class="pa-0 fill-height">
              <v-row no-gutters justify="end" align="center">
                <v-btn
                  v-if="enableLinkButton"
                  color="primary"
                  depressed
                  rounded
                  @click="openLinkNewTab"
                >
                  {{ btnText }}
                </v-btn>
                <v-spacer v-if="enableLinkButton" />
                <v-avatar v-for="item in techIntoArray" :key="item" size="48" tile class="mr-1 my-2">
                  <img :id="`${item}-img`" :src="matchLogoByName(item).src" :alt="matchLogoByName(item).alt">
                </v-avatar>
              </v-row>
            </v-container>
          </v-col>
        </slot>
      </v-row>
    </v-container>
  </v-card>
</template>
<style scoped>

</style>
<script>
export default {
  name: 'ProjectCard',
  props: {
    technologies: {
      type: String,
      default: ''
    },
    title: {
      type: String,
      required: true
    },
    description: {
      type: String,
      required: true
    },
    btnText: {
      type: String,
      default: 'Visitar'
    },
    linkButton: {
      type: String,
      default: ''
    },
    imgSrc: {
      type: String,
      default: 'https://cdn.vuetifyjs.com/images/cards/server-room.jpg'
    }

  },
  data () {
    return {
      languagesLogos: [
        {
          name: 'laravel',
          src: '/laravel-mark-rgb-red.svg',
          alt: 'Laravel Logo'
        },
        {
          name: 'mariadb',
          src: '/mariadb.svg',
          alt: 'MariaDB Logo'
        },
        {
          name: 'nodejs',
          src: '/Nodejs_logo.svg',
          alt: 'Node.js Logo'
        },
        {
          name: 'gitlab',
          src: '/gitlab.svg',
          alt: 'GitLab Logo'
        },
        {
          name: 'csharp',
          src: '/csharp_logo.svg',
          alt: 'C# Logo'
        },
        {
          name: 'github',
          src: '/github.svg',
          alt: 'Github Logo'
        },
        {
          name: 'mssql',
          src: '/mssql.svg',
          alt: 'Microsoft Sql Server Logo'
        },
        {
          name: 'vuejs',
          src: '/vuejs.svg',
          alt: 'VueJs Logo'
        },
        {
          name: 'nologo',
          src: '',
          alt: 'No logo'
        }
      ]
    }
  },
  computed: {
    enableLinkButton () {
      return this.linkButton.length > 0
    },
    techIntoArray () {
      return this.technologies.length > 0 ? this.technologies.split(',') : []
    }
  },
  methods: {
    matchLogoByName (val) {
      let result = this.languagesLogos.find(item => item.name === val.trim())
      if (!result)
        return this.languagesLogos.find(item => item.name === 'nologo')
      return result
    },
    openLinkNewTab () {
      window.open(this.linkButton, '_blank')
    }
  }
}
</script>
