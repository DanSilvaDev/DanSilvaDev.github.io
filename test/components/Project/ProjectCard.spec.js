import { shallowMount } from '@vue/test-utils'
import ProjectCard from '@/components/Project/ProjectCard'

const factory = () => shallowMount(ProjectCard, {
  propsData: {
    technologies: 'csharp,laravel',
    title: 'Test Title',
    description: 'Description Test',
    btnText: 'Check Page',
    linkButton: 'http://test.com',
    imgSrc: 'https://cdn.vuetifyjs.com/images/cards/server-room.jpg'
  }
})

const factoryWrongTech = () => shallowMount(ProjectCard, {
  propsData: {
    technologies: 'abc,xyz',
    title: 'Test Title',
    description: 'Description Test',
    btnText: 'Check Page',
    linkButton: 'http://test.com',
    imgSrc: 'https://cdn.vuetifyjs.com/images/cards/server-room.jpg'
  }
})

const factoryNoTech = () => shallowMount(ProjectCard, {
  propsData: {
    title: 'Test Title',
    description: 'Description Test',
    btnText: 'Check Page',
    linkButton: 'http://test.com',
    imgSrc: 'https://cdn.vuetifyjs.com/images/cards/server-room.jpg'
  }
})

const factoryNoButton = () => shallowMount(ProjectCard, {
  propsData: {
    technologies: 'csharp,laravel',
    title: 'Test Title',
    description: 'Description Test',
    btnText: 'Check Page',
    imgSrc: 'https://cdn.vuetifyjs.com/images/cards/server-room.jpg'
  }
})

describe('Project Card', () => {
  test('mounts properly', () => {
    const wrapper = factory()
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = factory()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('title shows correctly', () => {
    const wrapper = factory()
    expect(wrapper.find('h2').text()).toEqual('Test Title')
  })

  test('description shows correctly', () => {
    const wrapper = factory()
    expect(wrapper.find('p').text()).toEqual('Description Test')
  })

  test('button text set correctly', () => {
    const wrapper = factory()
    expect(wrapper.find('v-btn-stub').text()).toEqual('Check Page')
  })

  test('show correct specified technologies logo on tech input', () => {
    const wrapper = factory()
    expect(wrapper.find('v-btn-stub').exists()).toBeTruthy()
    expect(wrapper.find('v-spacer-stub').exists()).toBeTruthy()
    expect(wrapper.find('v-avatar-stub')).toBeTruthy()
    expect(wrapper.find('img')).toBeTruthy()
    expect(wrapper.find('#csharp-img').element.attributes.item(1).name).toEqual('src')
    expect(wrapper.find('#csharp-img').element.attributes.item(1).value).toEqual('/csharp_logo.svg')
    expect(wrapper.find('#csharp-img').element.attributes.item(2).name).toEqual('alt')
    expect(wrapper.find('#csharp-img').element.attributes.item(2).value).toEqual('C# Logo')
    expect(wrapper.find('#laravel-img').element.attributes.item(1).name).toEqual('src')
    expect(wrapper.find('#laravel-img').element.attributes.item(1).value).toEqual('/laravel-mark-rgb-red.svg')
    expect(wrapper.find('#laravel-img').element.attributes.item(2).name).toEqual('alt')
    expect(wrapper.find('#laravel-img').element.attributes.item(2).value).toEqual('Laravel Logo')
  })

  test('show no logo on wrongly technologies input', () => {
    const wrapper = factoryWrongTech()
    expect(wrapper.find('v-btn-stub').exists()).toBeTruthy()
    expect(wrapper.find('v-spacer-stub').exists()).toBeTruthy()
    expect(wrapper.find('v-avatar-stub')).toBeTruthy()
    expect(wrapper.find('img')).toBeTruthy()
    expect(wrapper.find('#abc-img').element.attributes.item(1).name).toEqual('src')
    expect(wrapper.find('#abc-img').element.attributes.item(1).value).toEqual('')
    expect(wrapper.find('#abc-img').element.attributes.item(2).name).toEqual('alt')
    expect(wrapper.find('#abc-img').element.attributes.item(2).value).toEqual('No logo')
    expect(wrapper.find('#xyz-img').element.attributes.item(1).name).toEqual('src')
    expect(wrapper.find('#xyz-img').element.attributes.item(1).value).toEqual('')
    expect(wrapper.find('#xyz-img').element.attributes.item(2).name).toEqual('alt')
    expect(wrapper.find('#xyz-img').element.attributes.item(2).value).toEqual('No logo')
  })

  test('show no image on empty technologies input', () => {
    const wrapper = factoryNoTech()
    expect(wrapper.find('v-avatar-stub').exists()).toBeFalsy()
    expect(wrapper.find('v-avatar-stub').exists()).toBeFalsy()
  })

  test('show no button on empty link input', () => {
    const wrapper = factoryNoButton()
    expect(wrapper.find('v-btn-stub').exists()).toBeFalsy()
    expect(wrapper.find('v-spacer-stub').exists()).toBeFalsy()
    expect(wrapper.find('v-avatar-stub').exists()).toBeTruthy()
    expect(wrapper.find('#csharp-img').exists()).toBeTruthy()
    expect(wrapper.find('#laravel-img').exists()).toBeTruthy()
  })

  test('calls openLinkNewTab on click', () => {
    window.open = jest.fn()
    const spyTst = jest.spyOn(ProjectCard.methods, 'openLinkNewTab')
    const wrapper = factory()
    wrapper.find('v-btn-stub').vm.$listeners.click() // trigger('click')
    expect(spyTst).toHaveBeenCalled()

    spyTst.mockReset()
    spyTst.mockRestore()
  })

  // test('clicked is true after click', () => {
  //   const wrapper = factory()
  //   wrapper.find('button').trigger('click')
  //   expect(wrapper.vm.clicked).toBe(true)
  // })
})
