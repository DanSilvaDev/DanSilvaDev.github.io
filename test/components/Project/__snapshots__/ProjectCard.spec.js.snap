// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Card renders properly 1`] = `
"<v-card-stub loaderheight=\\"4\\" tag=\\"div\\" outlined=\\"true\\">
  <v-container-stub tag=\\"div\\" fluid=\\"true\\" class=\\"fill-height\\">
    <v-row-stub tag=\\"div\\">
      <v-col-stub cols=\\"12\\" tag=\\"div\\" align=\\"center\\">
        <h2>Test Title</h2>
      </v-col-stub>
    </v-row-stub>
    <v-row-stub tag=\\"div\\" align=\\"center\\" justify=\\"center\\">
      <v-col-stub cols=\\"12\\" tag=\\"div\\" align=\\"center\\" justify=\\"center\\">
        <div class=\\"d-flex align-stretch\\">
          <v-img-stub height=\\"100%\\" maxheight=\\"150px\\" options=\\"[object Object]\\" position=\\"center center\\" src=\\"https://cdn.vuetifyjs.com/images/cards/server-room.jpg\\" transition=\\"fade-transition\\"></v-img-stub>
        </div>
      </v-col-stub>
      <v-col-stub cols=\\"12\\" tag=\\"div\\" align=\\"center\\">
        <p class=\\"text-justify\\">
          Description Test
        </p>
      </v-col-stub>
    </v-row-stub>
    <v-row-stub tag=\\"div\\">
      <v-divider-stub></v-divider-stub>
    </v-row-stub>
    <v-row-stub tag=\\"div\\">
      <v-col-stub cols=\\"12\\" tag=\\"div\\" align=\\"right\\">
        <v-container-stub tag=\\"div\\" fluid=\\"true\\" class=\\"pa-0 fill-height\\">
          <v-row-stub tag=\\"div\\" nogutters=\\"true\\" align=\\"center\\" justify=\\"end\\">
            <v-btn-stub color=\\"primary\\" rounded=\\"true\\" tag=\\"button\\" activeclass=\\"\\" depressed=\\"true\\" type=\\"button\\">
              Check Page
            </v-btn-stub>
            <v-spacer-stub></v-spacer-stub>
            <v-avatar-stub tile=\\"true\\" size=\\"48\\" class=\\"mr-1 my-2\\"><img id=\\"csharp-img\\" src=\\"/csharp_logo.svg\\" alt=\\"C# Logo\\"></v-avatar-stub>
            <v-avatar-stub tile=\\"true\\" size=\\"48\\" class=\\"mr-1 my-2\\"><img id=\\"laravel-img\\" src=\\"/laravel-mark-rgb-red.svg\\" alt=\\"Laravel Logo\\"></v-avatar-stub>
          </v-row-stub>
        </v-container-stub>
      </v-col-stub>
    </v-row-stub>
  </v-container-stub>
</v-card-stub>"
`;
